### PR TITLE
langchain[patch]: Enhance filter functionality for Elasticsearch VectorStore

### PR DESCRIPTION
**Pull request by LaurensTsestigh**
Based on an existing issue of the langchain project, I created a PR with the requested enhancement as we need this as well in our product where we use the langchain package: https://github.com/langchain-ai/langchainjs/issues/2699

As stated by the dosubot:

> With this modification, we could use any operator supported by Elasticsearch in our filters, not just the "term" operator.
> As for the different types of operators supported by Elasticsearch, they include but are not limited to: "term", "terms", "range", "exists", "missing", "prefix", "wildcard", "regexp", "fuzzy", "type", "ids", "geo_distance", "geo_bounding_box", "geo_polygon", "geo_distance_range", "geo_shape", and "script".

Thanks in advance

### CODE REVIEW

File: langchain/src/vectorstores/elasticsearch.ts [hunk 1]
```diff
@@ -38,7 +38,7 @@ export interface ElasticClientArgs {
 /**
  * Type representing a filter object in Elasticsearch.
  */
-type ElasticFilter = object;
+type ElasticFilter = { field: string, operator: string, value: any }[];
```
Comment (jacoblee93) [inline langchain/src/vectorstores/elasticsearch.ts:41]: This looks breaking? Shouldn't this allow `term` as a property?

Comment (LaurensTsestigh) [reply langchain/src/vectorstores/elasticsearch.ts:41]: Hi jacoblee, thanks for taking your time to check this PR out.

With the `operator` property we can be more flexible and allow for "term", "terms", "range", "exists", etc. instead.
As I typed it as an array of objects, we can have multiple different filters applied.
The breaking part seems to be in the docs, but it seems I can't change those?

Comment (jacoblee93) [reply langchain/src/vectorstores/elasticsearch.ts:41]: Ah got it, yeah looks like it might have just been a placeholder.

**Comment by jacoblee93** [comment]
Thanks! Can you merge and run `yarn format` and `yarn lint`

File: langchain/src/vectorstores/elasticsearch.ts [hunk 2]
```diff
@@ -299,11 +299,11 @@ export class ElasticVectorSearch extends VectorStore {
 
   private buildMetadataTerms(
     filter?: ElasticFilter
-  ): { term: Record<string, unknown> }[] {
+  ): { [operator: string]: { [field: string]: any } }[] {
     if (filter == null) return [];
     const result = [];
-    for (const [key, value] of Object.entries(filter)) {
-      result.push({ term: { [`metadata.${key}`]: value } });
+    for (const condition of filter) {
```
[the remark this answers is not in the document]
Comment (LaurensTsestigh) [reply langchain/src/vectorstores/elasticsearch.ts]: Yes you are right, we can't go breaking people's code. 
I've adapted it to be completely backwards-compatible

**Comment by jacoblee93** [comment]
Perfect, thank you!

**Comment by LaurensTsestigh** [comment]
> Perfect, thank you!

I've updated the PR with main and saw the dosubot removed the 'lgtm' label

**Comment by jacoblee93** [comment]
Ah, yeah if you push another commit I have to retrigger CI so please don't do that!

**Comment by LaurensTsestigh** [comment]
> Ah, yeah if you push another commit I have to retrigger CI so please don't do that unless there's a conflict!

I'll keep it in mind 😄 

Thanks a lot, have a nice day!

**Pull request merged**